### PR TITLE
Security Updates - July 2027

### PR DESCRIPTION
**Pull request by pdath**
# 📋 Pull Request Report: Security Updates July 2027

**Branch:** `security-updates-july-2027` → `main`  
**Status:** Ready for Review

---

## 📊 Overview

This branch implements **major security updates and dependency upgrades** to address vulnerabilities and modernize the codebase. Commits in this PR:

1. `be23302` - Upgraded to next.js 15
2. `a724723` - Upgraded eslint to 8.57.1
3. `193037b` - Upgraded to eslint V9
4. `bc79ea3` - Resolve eslint v9 issues
5. `a690a24` - Update modules to address security vulnerabilities

---

## 📦 Modified Files

| File | Changes |
|------|---------|
| `package.json` | Major version upgrades to Next.js, ESLint, Jest, TypeScript |
| `pnpm-lock.yaml` | Complete lockfile regeneration with new versions |

---

### ✅ High Priority

| Vulnerability | Package | Old Version | Fixed Version | CVE/GHSA |
|--------------|---------|-------------|--------------|----------|
| Prototype Pollution | lodash | Various | ^4.17.21 | CVE-2020-8203, etc. |
| Memory Exhaustion DoS | base-x | <4.0.1 | ^4.0.1 | GHSA-8597-w65w-8m6j |
| Command Injection | cross-spawn | <7.0.5 | ^7.0.5 | CVE-2024-21538 |

### 📋 Implementation Method
- Security pins added via `pnpm.overrides` in package.json
- All vulnerable transitive dependencies now use patched versions

---

### Core Framework
- **Next.js:** 14.x → **15.5.19**
  - App Router improvements
  - Performance optimizations
  - New features and bug fixes

- **React:** 18.3.1 → **18.3.1** (stable)
- **React DOM:** 18.3.1 → **18.3.1** (stable)

### Developer Tooling
- **ESLint:** 8.x → **9.39.4** → Migration to flat config format
- **TypeScript:** 5.x → **5.9.3**
- **Jest:** 29.x → **30.4.2**
- **Tailwind CSS:** 3.x → **3.4.19**

### Database & Backend
- **TypeORM:** 0.3.28 → **0.3.30**
- **pg:** 8.18.0 → **8.20.0**

### Testing
- **Testing Library:** React 16.3.2 → **16.3.2**
- **ts-jest:** 29.4.9 → **29.4.11**

---

### 1. ESLint v9 Migration
- **Change:** Flat config format (`eslint.config.js`) replaces `.eslintrc.*`
- **File Added:** `eslint.config.cjs` with full flat config implementation
- **Impact:** All ESLint rules now use new flat config format
- **Lint Issues Fixed:** 26 issues resolved (nullish coalescing, @ts-expect-error descriptions, prettier formatting)

### 2. Next.js 15 Upgrade
- **Note:** `serverExternalPackages` in next.config.js may need review
- **App Router:** Verify all pages and API routes work correctly
- **Edge Runtime:** Test edge functions if used

### 3. TypeScript 5.9.3
- **Stricter type checking** in some edge cases
- **New features:** Multiple configuration files, JSX improvements

---

### Lint Fixes Applied
- ✅ Replaced `||` with `??` (nullish coalescing) - 20+ occurrences
- ✅ Added descriptions to `@ts-expect-error` directives
- ✅ Fixed prettier formatting in jest.setup.js
- ✅ Removed unused imports (Worker in scripts/updateUsers.ts)
- ✅ Fixed return statement formatting (helpers.ts)

### Security
- ✅ Added pnpm overrides for vulnerable transitive dependencies

---

## 🧪 Testing Checklist

Before merging, please verify:

- [ ] `pnpm lint` passes without errors
- [ ] `pnpm build` completes successfully
- [ ] `pnpm dev` starts without errors
- [ ] `pnpm test` passes all tests
- [ ] Production deployment works correctly
- [ ] All pages load and function as expected
- [ ] API routes return correct responses

---

## 📈 Impact Summary

| Metric | Value |
|--------|-------|
| Security Vulnerabilities | 3 High severity - Addressed |
| Lint Errors | 26 - Fixed |
| Breaking Changes | 3 Major (ESLint, Next.js, TypeScript) |
| Dependency Updates | 30+ packages |
| Lockfile Version | Upgraded to 9.0 |

---

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved fallback handling for empty env/config values and HTTP/status text parsing across CKPool and database connections.
  * Updated user/worker pages to correctly await route params and handle missing worker names more precisely.
  * Refined error-message fallbacks, theme initialization/validation, and stats cleanup/counting for more accurate messaging and logging.

* **Chores**
  * Migrated linting to ESLint flat config and switched lint scripts to `eslint .`.
  * Ensured Jest loads `jest-dom` and tightened related TypeScript test expectations.

### CODE REVIEW

File: app/users/[address]/workers/[[...name]]/page.tsx [hunk 1]
```diff
@@ -73,7 +69,7 @@ export default async function WorkerPage({
           Back to User
         </Link>
         <h1 className="text-3xl font-bold text-accent">
-          {worker.name || <span className="italic">Unnamed Worker</span>}
+          {worker.name ?? <span className="italic">Unnamed Worker</span>}
```
[the remark this answers is not in the document]
Comment (pdath) [reply app/users/[address]/workers/[[...name]]/page.tsx:74]: I would rather it said "Unamed Worker" than a blank string.

File: lib/db_migration.ts [hunk 2]
```diff
@@ -9,11 +9,15 @@ import { WorkerStats } from './entities/WorkerStats';
...
-  username: process.env.DB_USER || 'postgres',
-  password: process.env.DB_PASSWORD || 'password',
-  database: process.env.DB_NAME || 'postgres',
+  // eslint-disable-next-line @typescript-eslint/prefer-nullish-coalescing
+  host: process.env.DB_HOST?.trim() || 'localhost',
+  // eslint-disable-next-line @typescript-eslint/prefer-nullish-coalescing
+  port: parseInt(process.env.DB_PORT?.trim() || '5432'),
+  // eslint-disable-next-line @typescript-eslint/prefer-nullish-coalescing
+  username: process.env.DB_USER?.trim() || 'postgres',
+  password: process.env.DB_PASSWORD ?? 'password',
```
[the remark this answers is not in the document]
Comment (pdath) [reply lib/db_migration.ts:18]: DB_USER and DB_PASSWORD are optional when using a Unix socket.  When using a Unix socket the username and password are ignored.

File: lib/db.ts [hunk 3]
```diff
@@ -16,11 +16,15 @@ import { WorkerStats } from './entities/WorkerStats';
...
-  username: process.env.DB_USER || 'postgres',
-  password: process.env.DB_PASSWORD || 'password',
-  database: process.env.DB_NAME || 'postgres',
+  // eslint-disable-next-line @typescript-eslint/prefer-nullish-coalescing
+  host: process.env.DB_HOST?.trim() || 'localhost',
+  // eslint-disable-next-line @typescript-eslint/prefer-nullish-coalescing
+  port: parseInt(process.env.DB_PORT?.trim() || '5432'),
+  // eslint-disable-next-line @typescript-eslint/prefer-nullish-coalescing
+  username: process.env.DB_USER?.trim() || 'postgres',
+  password: process.env.DB_PASSWORD ?? 'password',
```
[the remark this answers is not in the document]
Comment (pdath) [reply lib/db.ts:25]: DB_USER and DB_PASSWORD are optional when using a Unix socket.  When using a Unix socket the username and password are ignored.

**Review by Ari4ka — APPROVED** [review]
